hoverkite-protocol/src/lib.rs

```diff
@@ -87,81 +87,67 @@ impl Command {
     }
 
     pub fn parse(buf: &[u8]) -> nb::Result<Self, ParseError> {
-        let first = buf.get(0).ok_or(WouldBlock)?;
-        match first {
-            b'l' => Ok(Command::SetSideLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'o' => Ok(Command::SetOrangeLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'r' => Ok(Command::SetRedLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'g' => Ok(Command::SetGreenLed(ascii_to_bool(
-                *buf.get(1).ok_or(WouldBlock)?,
-            )?)),
-            b'b' => Ok(Self::ReportBattery),
-            b'c' => Ok(Self::ReportCharger),
-            b'S' => {
-                if buf.len() < 5 {
-                    return Err(WouldBlock);
-                }
-                let min_power = i16::from_le_bytes(buf[1..3].try_into().unwrap());
-                let max_power = i16::from_le_bytes(buf[3..5].try_into().unwrap());
-
-                Ok(Self::SetMaxSpeed(min_power..=max_power))
+        let command = match *buf {
+            [b'l', on] => Self::SetSideLed(ascii_to_bool(on)?),
+            [b'o', on] => Self::SetOrangeLed(ascii_to_bool(on)?),
+            [b'r', on] => Self::SetRedLed(ascii_to_bool(on)?),
+            [b'g', on] => Self::SetGreenLed(ascii_to_bool(on)?),
+            [b'b'] => Self::ReportBattery,
+            [b'c'] => Self::ReportCharger,
+            [b'S', min_lsb, min_msb, max_lsb, max_msb] => {
+                let min_power = i16::from_le_bytes([min_lsb, min_msb]);
+                let max_power = i16::from_le_bytes([max_lsb, max_msb]);
+                Self::SetMaxSpeed(min_power..=max_power)
             }
-            b'K' => {
-                if buf.len() < 3 {
-                    return Err(WouldBlock);
-                }
-                let spring = u16::from_le_bytes(buf[1..3].try_into().unwrap()).into();
-                Ok(Self::SetSpringConstant(spring))
+            [b'K', lsb, msb] => {
+                let spring = u16::from_le_bytes([lsb, msb]);
+                Self::SetSpringConstant(spring)
             }
-            b'n' => Ok(Self::RemoveTarget),
-            b'T' => {
-                if buf.len() < 9 {
-                    return Err(WouldBlock);
-                }
-                let target = i64::from_le_bytes(buf[1..9].try_into().unwrap());
-                Ok(Self::SetTarget(target))
+            [b'n'] => Self::RemoveTarget,
+            [b'T', b0, b1, b2, b3, b4, b5, b6, b7] => {
+                let target = i64::from_le_bytes([b0, b1, b2, b3, b4, b5, b6, b7]);
+                Self::SetTarget(target)
             }
-            b'e' => Ok(Self::Recenter),
-            b'+' => Ok(Self::IncrementTarget),
-            b'-' => Ok(Self::DecrementTarget),
-            b'p' => Ok(Self::PowerOff),
-            _ => Err(Other(ParseError)),
-        }
+            [b'e'] => Self::Recenter,
```

Comment:
Maybe it's nicer to write this as `b"e"`.

Reply:
I quite like the symmetry of having them all as array/slice patterns.